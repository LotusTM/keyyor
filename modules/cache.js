const LRUCache = require('lru-cache')
const hash = require('hash-sum')

const cache = new LRUCache()

/**
 * Cache function into specified cache key or based on hash of its arguments
 * @param  {function} func       Function to be cached
 * @param  {string}   [cacheKey] Optional cache store key to store result.
 *                               If store already has `cacheKey` value, it will be
 *                               immediately returned
 * @return {function} Cached function to be called with desired arguments
 * @example
 *  cacheFunc(renderFunc)('{{ value }}', { value: 'test' })
 *  cacheFunc(renderFunc, 'matter')('{{ value }}', { value: 'test2' })
 */
const cacheFunc = (func, cacheKey) => (...args) => {
  const key = cacheKey || hash(args)
  const cached = cache.get(key)

  if (cached) return cached

  const result = func(...args)

  cache.set(key, result)

  return result
}

module.exports = {
  cache,
  cacheFunc
}
