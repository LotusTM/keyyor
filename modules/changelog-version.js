const { exec } = require('child_process')
const { readFile, writeFileSync } = require('fs')
const { join } = require('path')
const moment = require('moment')
const pkg = require(join(process.cwd(), 'package.json'))
const { green, red } = require('chalk')

const isFromCLI = require.main === module
const logError = (message) => console.error(red(message))

const VERSION = pkg.version
const CHANGELOG_FILENAME = 'CHANGELOG.md'
const REPOSITORY_URL = pkg.repository.url
const DATE = moment().format('YYYY-MM-DD')
const HEAD_PATTERN = /^## \[?(HEAD|Unreleased)\]?.+$/m
const VERSION_PATTERN = /^## \[?(\d+\.\d+\.\d+)/m

/**
 * Update inputted changelog version
 *
 * @param  {object} [options]                   Options
 * @param  {string} [options.input]             Input to be changed
 * @param  {string} [options.version]           Version to be used
 * @param  {string} [options.repositoryURL]     URL to the project repository
 * @param  {string} [options.date]              Version date
 * @param  {RegExp} [options.headPattern]       Pattern to match HEAD heading to update
 * @param  {RegExp} [options.versionPattern]    Pattern to match versions in changelog
 * @return {string} Updated input
 */
const updateVersion = ({
  input,
  version = VERSION,
  repositoryURL = REPOSITORY_URL,
  date = DATE,
  headPattern = HEAD_PATTERN,
  versionPattern = VERSION_PATTERN
} = {}) => {
  if (!version) {
    logError('[changelog-version] please, specify `package.json` `version` property')
    process.exit(1)
  }

  if (!repositoryURL || !repositoryURL.includes('github.com')) {
    logError('[changelog-version] please, specify `package.json` `repository.url` property with a valid Github URL')
    process.exit(1)
  }

  if (repositoryURL.endsWith('.git')) {
    repositoryURL = repositoryURL.slice(0, -4)
  }

  const hasHead = headPattern.test(input)

  if (!hasHead) {
    logError('[changelog-version] changelog does not have HEAD version, can not update it with latest version')
    process.exit(1)
  }

  const previousVersion = input.match(versionPattern)[1]
  const releaseHeader = `## [${version}](${repositoryURL}/compare/v${previousVersion}...v${version}) - ${date}`
  const newHEADHeader = `## [HEAD](${repositoryURL}/compare/v${version}...HEAD)`

  return input.replace(headPattern, `${newHEADHeader}\n\n${releaseHeader}`)
}

/**
 * Update `CHANGELOG.md` HEAD to the latest version,
 * with proper compare URL and date, as well as
 * add new `HEAD` header with proper URL.
 *
 * CHANGELOG should follow [keepchangelog.com specs](https://keepachangelog.com).
 *
 * Relies on `package.json` `version` and `repository.url` values.
 *
 * Sould be run in `version` NPM script.
 *
 * @param  {object} [options]                   Options
 * @param  {string} [options.version]           Version to be used
 * @param  {string} [options.changelogFilename] Changelog filename
 * @param  {string} [options.repositoryURL]     URL to the project repository
 * @param  {string} [options.date]              Version date
 * @param  {RegExp} [options.headPattern]       Pattern to match HEAD heading to update
 * @param  {RegExp} [options.versionPattern]    Pattern to match versions in changelog
 * @return {void} It just updates `CHANGELOG.md`
 */
const updateChangelogVersion = ({
  version = VERSION,
  changelogFilename = CHANGELOG_FILENAME,
  repositoryURL = REPOSITORY_URL,
  date = DATE,
  headPattern = HEAD_PATTERN,
  versionPattern = VERSION_PATTERN
} = {}) => readFile(changelogFilename, { encoding: 'utf-8' }, (error, changelog) => {
  if (error) return console.log(`[changelog-version] seems to be no ${changelogFilename}, skipping changelog version update`)

  writeFileSync(changelogFilename, updateVersion({
    input: changelog,
    version,
    changelogFilename,
    repositoryURL,
    date,
    headPattern,
    versionPattern
  }))

  exec(`git add ${changelogFilename}`, (error, stdout) => {
    if (error) {
      logError('[changelog-version] An error occurred while staging updated changelog:\n')
      logError(error)
      process.exit(1)
    }

    console.log(green(`\n[changelog-version] succesfully updated changelog to the v${version}`))
  })
})

if (isFromCLI) updateChangelogVersion()

module.exports = {
  updateVersion,
  updateChangelogVersion
}
