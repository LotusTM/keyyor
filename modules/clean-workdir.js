const { exec, execSync } = require('child_process')
const { green, red } = require('chalk')

const isFromCLI = require.main === module
const logError = (message) => console.error(red(message))

const checkIsWorkDirClean = () => {
  exec('git --version', (error) => {
    if (error) {
      console.log('[clean-workdir] No Git has been detected.\n\nSkipping clean working directory test.\n')
      return console.log(error)
    }

    exec('git status --porcelain', (error, stdout) => {
      if (error) {
        logError('[clean-workdir] an error occured:\n')
        return logError(error)
      }

      if (stdout) {
        logError('[clean-workdir] Modified files have been detected.\n')
        logError('Build task should not generate or change files outside the build folder:\n')
        logError(stdout)
        console.error(execSync('git diff --color').toString())
        return process.exit(1)
      }

      console.log(green('[clean-workdir] no modified files detected outside of the build folder. Live long and prosper'))
    })
  })
}

if (isFromCLI) checkIsWorkDirClean()

module.exports = checkIsWorkDirClean
