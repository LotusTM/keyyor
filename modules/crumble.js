const _ = require('lodash')
const { extname } = require('path')

/**
 * Explodes string path into array breadcrumb
 * @param  {string} to Relative or absolute path
 * @return {array}     Array, which consists of path's fragments
 * @example crumble('/url/to/index.html') -> `['url', 'to']`
 * @example crumble('/url/to/404.html') -> `['url', 'to', '404']`
 * @example crumble('/') -> `['index']`
*/
module.exports = (to) => {
  if (to === '/') return ['index']

  const breadcrumb = _.chain(to)
    .trimStart('/')
    .trimEnd('/')
    .replace(new RegExp(`${extname(to)}$`, 'ig'), '')
    .split('/')
    .value()

  if ((breadcrumb.length >= 2) && (_.last(breadcrumb) === 'index')) {
    breadcrumb.pop()
  }

  return breadcrumb
}
