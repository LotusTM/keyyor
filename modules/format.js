const { sprintf } = require('sprintf-js')
const { vsprintf } = require('sprintf-js')
const traverse = require('./traverse')

/**
 * Replace placeholders with provided values via `sprintf` or `vsprintf`.
 * `printf` type will be chosen depending on provided placeholders
 * @param {*}                   input           Input in which should be made replacement
 * @param {string|object|array} placeholders... List of placeholders, object with named
 *                                              placeholders or arrays of placeholders
 * @return {*} Transformed input
*/
module.exports = (input, ...placeholders) => {
  const [placeholder] = placeholders
  const useVsprintf = placeholders.length === 1 && Array.isArray(placeholder)

  return traverse(
    input,
    (tmpl) => useVsprintf ? vsprintf(tmpl, placeholder) : sprintf(tmpl, ...placeholders),
    (string) => /%([\w+-]|\.\d|[\w_]'|\([\w])/.test(string)
  )
}
