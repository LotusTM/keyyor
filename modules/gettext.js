const Gettext = require('node-gettext')
const { po } = require('gettext-parser')
const { join, extname } = require('path')
const { readFileSync } = require('fs')
const { file: { expand } } = require('grunt')

const DEFAULT_DOMAIN = 'messages'

const getLocaleDirs = (cwd) => expand({ cwd, filter: 'isDirectory' }, '*', '!*templates')

const readPoMessages = (path) => {
  const file = readFileSync(path)
  return po.parse(file)
}

const autobindTextdomains = (gt, cwd, locale) => {
  cwd = join(cwd, locale)

  expand({ cwd, filter: 'isFile' }, '{,**/}*.{mo,po}')
    .forEach(domainpath => {
      const ignored = new RegExp(`${extname(domainpath)}$|LC_MESSAGES\\/`)
      const domain = domainpath.replace(ignored, '')
      const domainMessages = readPoMessages(join(cwd, domainpath))

      gt.addTranslations(locale, domain, domainMessages)
    })
}

module.exports = (cwd) => {
  if (!cwd) {
    throw new Error('[gettext] argument `options.cwd` should be privided')
  }

  const gt = new Gettext()

  getLocaleDirs(cwd).forEach((localeDir) => {
    autobindTextdomains(gt, cwd, localeDir)
  })

  gt.on('error', (error) => console.error(
    `[gettext] ${error}`
  ))

  return {
    setLocale: gt.setLocale,
    setTextdomain: gt.setTextDomain,
    on: gt.on,
    off: gt.off,
    gettext: gt.gettext,
    dgettext: gt.dgettext,
    ngettext: gt.ngettext,
    dngettext: gt.dngettext,
    pgettext: gt.pgettext,
    dpgettext: gt.dpgettext,
    npgettext: gt.npgettext,
    dnpgettext: gt.dnpgettext,

    nunjucksExtensions (env, currentLocale) {
      gt.setLocale(currentLocale)
      gt.setTextDomain(DEFAULT_DOMAIN)

      env.addGlobal('setLocale', (locale = currentLocale) => gt.setLocale.bind(gt)(locale))
      env.addGlobal('setTextdomain', (domain = DEFAULT_DOMAIN) => gt.setTextDomain.bind(gt)(domain))
      env.addGlobal('gettext', gt.gettext.bind(gt))
      env.addGlobal('dgettext', gt.dgettext.bind(gt))
      env.addGlobal('ngettext', gt.ngettext.bind(gt))
      env.addGlobal('dngettext', gt.dngettext.bind(gt))
      env.addGlobal('pgettext', gt.pgettext.bind(gt))
      env.addGlobal('dpgettext', gt.dpgettext.bind(gt))
      env.addGlobal('npgettext', gt.npgettext.bind(gt))
      env.addGlobal('dnpgettext', gt.dnpgettext.bind(gt))
    }
  }
}
