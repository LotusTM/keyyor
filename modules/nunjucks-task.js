const { get, merge } = require('lodash')
const { join } = require('path')
const crumble = require('./crumble')
const humanReadableUrl = require('./humanReadableUrl')
const i18nTools = require('./i18n-tools')
const nunjucksExtensions = require('./nunjucks-extensions')
const render = require('./nunjucks-render')

const prepareMatter = (env, data) => {
  const { matter } = data.SITE
  const theMatter = typeof matter === 'function' ? matter() : matter
  const rendered = render(env, data, theMatter)

  data.SITE.matter = rendered
  data.SITE.matter.$raw = theMatter
}

module.exports = function (config) {
  config = merge({
    options: {
      data: {},
      humanReadableUrls: false,
      humanReadableUrlsExclude: /^(index|\d{3})$/
    }
  }, config)

  const { files } = config
  let { configureEnvironment, preprocessData, humanReadableUrls, humanReadableUrlsExclude, currentLocale, locales, baseLocale, gettext } = config.options
  const { getLocaleProps, getLocaleDir, getLangcode, getRegioncode, isoLocale } = i18nTools

  currentLocale = currentLocale || baseLocale

  if (!baseLocale && (typeof baseLocale !== 'string')) {
    throw new Error('[nunjucks-task] base locale should be specified as `options.baseLocale` string')
  }

  if (!locales && (typeof locales !== 'object')) {
    throw new Error('[nunjucks-task] locales properties should be specified as `options.locales` object')
  }

  if (!gettext && (typeof gettext !== 'object')) {
    throw new Error('[nunjucks-task] gettext instance should be provided as `options.gettext` object')
  }

  if (!files) {
    throw new Error('[nunjucks-task] `src` and `dest` should be provided as array of objects with `expand: true`')
  }

  const localeProps = getLocaleProps(locales, currentLocale)
  const localeDir = getLocaleDir(locales, currentLocale)

  config = merge(config, {
    options: {
      configureEnvironment (env, nunjucks) {
        nunjucksExtensions(env)
        gettext.nunjucksExtensions(env, currentLocale)
        i18nTools.nunjucksExtensions(env)

        if (typeof configureEnvironment === 'function') {
          configureEnvironment.call(this, env, nunjucks)
        }

        const { data } = env.opts
        const { SITE } = data
        const { matter, images } = SITE

        if ((typeof matter !== 'function') && (typeof matter !== 'object')) {
          throw new Error(`[nunjucks-task] \`options.data.SITE.matter\` should be a function, which returns matter object, or a plain matter object, ${typeof matter} provided`)
        }

        if ((typeof images !== 'function') && (typeof images !== 'object')) {
          throw new Error(`[nunjucks-task] \`options.data.SITE.images\` should be a function, which returns matter object, or a plain matter object, ${typeof images} provided`)
        }

        prepareMatter(env, data)
        SITE.images = typeof images === 'function' ? images() : images
      },

      preprocessData (data) {
        const pagepath = humanReadableUrl(this.src[0].replace((this.orig.cwd || this.orig.orig.cwd), ''), humanReadableUrlsExclude)
        const breadcrumb = crumble(pagepath)
        const { matter } = data.SITE
        const { props } = get(matter, breadcrumb)

        data.PAGE = merge(data.PAGE, {
          props: {
            locale: currentLocale,
            isoLocale: isoLocale(currentLocale),
            language: getLangcode(currentLocale),
            region: getRegioncode(currentLocale),
            rtl: localeProps.rtl
          }
        }, { props })

        if (typeof preprocessData === 'function') {
          data = preprocessData.call(this, data)
        }

        return data
      }
    }
  })

  files.map((file) => {
    if (!file.expand) {
      throw new Error('[nunjucks-task] files mapping should use `expand: true`')
    }

    const { rename } = file

    file.rename = (dest, src) => {
      const newSrc = humanReadableUrls ? humanReadableUrl(src, humanReadableUrlsExclude) : src
      const newDest = join(dest, localeDir)

      if (typeof rename === 'function') {
        return rename.call(this, newDest, newSrc)
      }

      return join(newDest, newSrc)
    }
  })

  return config
}

module.exports.prepareMatter = prepareMatter
