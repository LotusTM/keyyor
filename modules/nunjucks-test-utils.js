const nunjucks = require('nunjucks')
const cloneDeep = require('lodash/cloneDeep')
const { grunt } = require('./grunt')
const nunjucksExtensions = require('./nunjucks-extensions')
const i18nTools = require('./i18n-tools')
const { cache } = require('./cache')

const env = nunjucks.configure(grunt.config('path.source.templates'))

nunjucksExtensions(env)
i18nTools.nunjucksExtensions(env)

/**
 * Utility function for testing Nunjucks template
 * Renders template with specific context and usual
 * for Kotsu environment
 * @param  {string}  template          Template to render
 * @param  {object}  [context]         Template context
 * @param  {boolean} [parse]           Parse output with `JSON.parse`.
 *                                     Useful for object or array dumps
 * @param  {boolean} [resetCache=true] Reset cache
 * @return {string} Rendered template
 */
const renderString = (template, context, parse, resetCache = true) => {
  if (resetCache) cache.reset()

  const rendered = env.renderString(template, cloneDeep(context))

  if (parse) {
    return JSON.parse(rendered)
  }

  return rendered
}

module.exports = { env, renderString }
