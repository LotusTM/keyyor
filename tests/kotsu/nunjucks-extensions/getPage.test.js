/* eslint-env jest */

import { env, renderString } from '../../../modules/nunjucks-test-utils'
import { prepareMatter } from '../../../modules/nunjucks-task'

const render = (template, context = mockContext, parse) => renderString(template, context, parse)
const mockContext = {
  SITE: {
    name: 'Test name',
    matter: {
      'index': {
        'props': { 'url': '/' }
      },
      'blog': {
        'props': { 'url': '/blog' },
        '2015-10-12-example-article': {
          'props': { 'url': '/blog/2015-10-12-example-article' }
        }
      },
      'contexVar': {
        'props': { 'var': '__contextvar value: {{ __contextvar }}' }
      },
      'globalVar': {
        'props': { 'var': 'SITE.name value: {{ SITE.name }}' }
      },
      'globalFunc': {
        'props': { 'func': 'Crumbled url: {{ crumble("blog") }}' }
      }
      // @todo Disabled, because it will always fail due to endless recursion in cached `getPage`
      // 'testGetPage': {
      //   'props': { 'func': 'Blog url: {{ getPage("blog").props.url }}' }
      // }
    }
  }
}

describe('Nunjucks global function `getPage()`', () => {
  beforeAll(() => prepareMatter(env, mockContext))

  describe('should get from', () => {
    it('root string-based path', () => {
      expect(render(`{{ getPage('blog')|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root', () => {
      expect(render(`{{ getPage('/').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root string-based path with leading slash', () => {
      expect(render(`{{ getPage('/blog')|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root string-based path with trailing slash', () => {
      expect(render(`{{ getPage('blog/')|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root string-based path with leading and trailing slashes', () => {
      expect(render(`{{ getPage('/blog/')|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root Array-based path', () => {
      expect(render(`{{ getPage(['blog'])|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested dot-notation path', () => {
      expect(render(`{{ getPage('blog.2015-10-12-example-article').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested Array-based path', () => {
      expect(render(`{{ getPage(['blog', '2015-10-12-example-article']).props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested url-like path', () => {
      expect(render(`{{ getPage('blog/2015-10-12-example-article').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested url-like path with leading slash', () => {
      expect(render(`{{ getPage('/blog/2015-10-12-example-article').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested url-like path with trailing slash', () => {
      expect(render(`{{ getPage('blog/2015-10-12-example-article/').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested url-like path with leading and trailing slash', () => {
      expect(render(`{{ getPage('/blog/2015-10-12-example-article/').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })
  })

  it('should get $raw data', () => {
    expect(render(`{{ getPage('contexVar').$raw|dump|safe }}`, undefined, true)).toMatchSnapshot()
  })

  it('should allow to render $raw data', () => {
    expect(render(`
      {% set __contextvar = 'testing __contextvar value' %}
      {{ getPage('contexVar').$raw.props.var|render() }}
    `)).toMatchSnapshot()
  })

  describe('should get rendered data', () => {
    it('with global context', () => {
      expect(render(`{{ getPage('globalVar').props.var }}`)).toMatchSnapshot()
    })

    it('with global custom functions', () => {
      expect(render(`{{ getPage('globalFunc').props.func }}`)).toMatchSnapshot()
    })

    it('and ignore current context', () => {
      expect(render(`
        {% set __contextvar = 'testing __contextvar value' %}
        {{ getPage('contexVar').props.var }}
      `)).toMatchSnapshot()
    })
  })
})
