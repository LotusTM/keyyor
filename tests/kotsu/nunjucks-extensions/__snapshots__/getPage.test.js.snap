// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nunjucks global function \`getPage()\` should allow to render $raw data 1`] = `
"
      
      __contextvar value: testing __contextvar value
    "
`;

exports[`Nunjucks global function \`getPage()\` should get $raw data 1`] = `
Object {
  "props": Object {
    "var": "__contextvar value: {{ __contextvar }}",
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested Array-based path 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested dot-notation path 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested url-like path 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested url-like path with leading and trailing slash 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested url-like path with leading slash 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested url-like path with trailing slash 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root 1`] = `
Object {
  "url": "/",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root Array-based path 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root string-based path 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root string-based path with leading and trailing slashes 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root string-based path with leading slash 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root string-based path with trailing slash 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get rendered data and ignore current context 1`] = `
"
        
        __contextvar value: 
      "
`;

exports[`Nunjucks global function \`getPage()\` should get rendered data with global context 1`] = `"SITE.name value: Test name"`;

exports[`Nunjucks global function \`getPage()\` should get rendered data with global custom functions 1`] = `"Crumbled url: blog"`;
